network/src/protocols/network/mod.rs

```diff
@@ -79,51 +76,68 @@ impl<TMessage: PartialEq> PartialEq for Event<TMessage> {
     }
 }
 
-/// Configuration needed for AptosNet applications to register with the network
-/// builder. Supports client-only, service-only, and P2p (both) applications.
-// TODO(philiphayes): separate configs for client & server?
-#[derive(Clone, Default)]
-pub struct NetworkApplicationConfig {
-    /// The set of protocols needed for this application.
-    pub protocols: ProtocolIdSet,
-    /// The config for the inbound message queue from network to the application.
-    /// Used for specifying the queue style (e.g. FIFO vs LIFO) and sub-queue max
-    /// capacity.
-    // TODO(philiphayes): only relevant for services
-    // TODO(philiphayes): in the future, use a Service trait here instead?
-    pub inbound_queue: Option<aptos_channel::Config>,
+/// Configuration needed for the client side of AptosNet applications
+#[derive(Clone)]
+pub struct NetworkClientConfig {
+    /// Direct send protocols for the application (sorted by preference, highest to lowest)
+    pub direct_send_protocols_and_preferences: Vec<ProtocolId>,
+    /// RPC protocols for the application (sorted by preference, highest to lowest)
+    pub rpc_protocols_and_preferences: Vec<ProtocolId>,
 }
 
-impl NetworkApplicationConfig {
-    /// AptosNet client configuration. Requires the set of protocols used by the
-    /// client in its requests.
-    pub fn client(protocols: impl IntoIterator<Item = ProtocolId>) -> Self {
+impl NetworkClientConfig {
+    pub fn new(
+        direct_send_protocols_and_preferences: Vec<ProtocolId>,
+        rpc_protocols_and_preferences: Vec<ProtocolId>,
+    ) -> Self {
         Self {
-            protocols: ProtocolIdSet::from_iter(protocols),
-            inbound_queue: None,
+            direct_send_protocols_and_preferences,
+            rpc_protocols_and_preferences,
         }
     }
+}
+
+/// Configuration needed for the service side of AptosNet applications
+#[derive(Clone)]
+pub struct NetworkServiceConfig {
+    /// Direct send protocols for the application (sorted by preference, highest to lowest)
+    pub direct_send_protocols_and_preferences: Vec<ProtocolId>,
```

Comment:
Hmm.. with this change, we need to duplicate protocol_ids both on the client and server configuration - is my understanding correct? Can you explain what happens if there is a configuration mismatch?

Reply:
Yes, but I think this is okay, for a couple reasons: (i) protocols are actually negotiated between client and server, e.g., there is an order preference that is done when messages are actually sent. See here (https://github.com/aptos-labs/aptos-core/blob/main/network/src/application/interface.rs#L112); (ii) we only wire this up once, using the same constants inside each ApplicationNetworkConfig, so there shouldn't be a case where they disagree within a single node. We can always add some tests around this (to be extra sure no-one misconfigures a node); and (iii) I can see future use cases where we actually want this to be the case (e.g., to maintain backward compatibility when rolling out new services, etc.)